source/credentials_provider_imds.c

```diff
@@ -102,10 +103,12 @@ struct aws_credentials_provider_imds_user_data {
     bool token_required;
     uint8_t retry_count;
     int status_code;
+
+    struct aws_atomic_var ref_count;
+    bool callback_invoked;
```

Comment:
This is to prevent callback from being invoked multiple times, currently there are two states related to the same state function: s_query_error